### ISSUE TITLE
FYI:  two things that I've done with/to PhpParser that might be useful permanent features

### ISSUE THREAD

**Issue by SundialServices** [nikic/PHP-Parser#272]
I am using PhpParser _(thank you! thank you!! thank you!!!)_ to get a grip on an 11-year old application that consists of more than 4,200 PHP files.  (Thousands of includes, hundreds of defines, you name it.)  To that end, I developed two modifications to PhpParser that might be useful optional features:

(1)  A "depth-first" traverser, which immediately recurses into any array that is returned (in either of the two cases) from an "after-visit" handler.  I fully understand that not every visitor would be compatible with this behavior, but it's vital for a special visitor I've written to handle includes.  (These let me fully-expand the AST, prior to subsequent traverses of that expanded AST.)

(2)  The option to include "parent," "prev," "next," and "children" attributes into a node as it is traversed.  In practice, you run a traverser (with no visitors) with these options turned on, to populate these attributes, and then you can "walk the tree" generically.

These are "extremely special-purpose" features, but very useful to me.  Maybe they would be useful to other people, as well.  They certainly are straightforward modifications to the existing code, and they would have zero-impact to what you have now when turned-off.  (And, both of these _would be_ "things that are normally turned-off" ...)

**Comment by nikic** [comment]
For case (2), is there any problem implementing this in a visitor rather than the traverser itself?

**Comment by felixfbecker** [comment]
@nikic I do find myself often in the position where I would like to pass a Node object around in the application, but then need to get access to it's parent node. For (a real world) example, say I want to find out the definition of symbol at a position. I have a generic function that returns a node at a position. Now I have a node, lets say it is a variable access. To find the declaration, I need to walk the tree up to the closest function/method node and then search inside the parameters. Atm you can always access the children of a node but you cannot access the parent, like you can in a DOM node for example.

**Comment by SundialServices** [comment]
I'd like to add one more note to my point (2) above:  in order to create parent/sibling branches without creating storage-loops (such that *e.g. `print_r()` never completes!),* I used integers for these attributes, which are indexes to a separately-stored array of node-references.  Each node also carries an attribute which contains its own index-number.  (And yes, my ever-suspicious routines confirm that the two are the same.)

So the `->nextSibling` is not an attribute, but a method.  Works the same way but doesn't mess with the PHP garbage-collector's mind.

Recalling now how I did the code, *I* found it more expedient to make slight modifications to the runner itself, because it most-easily had the necessary information, whereas a visitor sees one node at a time.  I'm sure that it could be done both ways.  In any case, I'd like to see this as an official *(and, optional ...)* feature.

With regard to my point (1) above, I wound up co-opting the `HaltCompiler` node type ... which didn't seem to be used.  To this node-type is added an indicator of what include-file you are entering, through mandatory attributes.  By detecting this node type, and confirming that it does bear these attributes *(vs. being an actual instruction to halt the compiler!),* one can traverse the entire include-file structure and leave indicators within the AST of where each one begins.
